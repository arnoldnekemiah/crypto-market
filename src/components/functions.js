import { NavLink, Route } from 'react-router-dom';
import Token from './Token';

export const generateNavLinks = (cryptoArray) => {
  const linkArray = [];
  for (let rank = 1; rank <= cryptoArray.length; rank += 1) {
    linkArray.push(
      <li key={rank} className="gridItem ">
        <NavLink className="link row gridLink" to={`/${rank}`}>
          <h2>{`#${cryptoArray[rank - 1].rank}`}</h2>
          <p>{`${cryptoArray[rank - 1].name}`}</p>
        </NavLink>
      </li>,
    );
  }
  return linkArray;
};

export const filterByRankOrName = (arr, parameter) => {
  const argument = parameter.toLowerCase().trim();
  if (!argument) {
    return []; // Return an empty array if the query is empty.
  }

  const filteredTokens = arr.filter(
    (coin) => coin.rank === argument || coin.name.toLowerCase().includes(argument),
  );

  return filteredTokens;
};

export const generateRoutes = (arr) => {
  const routes = [];

  for (let rank = 1; rank <= arr.length; rank += 1) {
    routes.push(
      <Route
        key={rank}
        path={`/${rank}`}
        element={<Token rank={`${rank.toString()}`} />}
      />,
    );
  }

  return routes;
};
