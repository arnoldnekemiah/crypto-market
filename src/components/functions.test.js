import React from 'react';
import { BrowserRouter as Router, Routes } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { generateNavLinks, filterByRankOrName, generateRoutes } from './functions';

describe('generateNavLinks', () => {
  it('generates navigation links correctly', () => {
    const cryptoArray = [
      { rank: 1, name: 'Bitcoin' },
      { rank: 2, name: 'Ethereum' },
    ];

    const navLinks = generateNavLinks(cryptoArray);

    const component = renderer.create(
      <Router>
        <ul>{navLinks}</ul>
      </Router>,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('filterByRankOrName', () => {
  it('filters tokens by rank or name', () => {
    const cryptoArray = [
      { rank: 1, name: 'Bitcoin' },
      { rank: 2, name: 'Ethereum' },
    ];

    const filteredTokens = filterByRankOrName(cryptoArray, 'bitcoin');

    expect(filteredTokens).toMatchSnapshot();
  });
});

describe('generateRoutes', () => {
  it('generates routes correctly', () => {
    const cryptoArray = [
      { rank: 1, name: 'Bitcoin' },
      { rank: 2, name: 'Ethereum' },
    ];

    const routes = generateRoutes(cryptoArray);

    const component = renderer.create(
      <Router>
        <Routes>
          {routes}
        </Routes>
      </Router>,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
