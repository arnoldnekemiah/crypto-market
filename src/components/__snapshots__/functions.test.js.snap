// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterByRankOrName filters tokens by rank or name 1`] = `
Array [
  Object {
    "name": "Bitcoin",
    "rank": 1,
  },
]
`;

exports[`generateNavLinks generates navigation links correctly 1`] = `
<ul>
  <li
    className="gridItem "
  >
    <a
      className="link row gridLink"
      href="/1"
      onClick={[Function]}
    >
      <h2>
        #1
      </h2>
      <p>
        Bitcoin
      </p>
    </a>
  </li>
  <li
    className="gridItem "
  >
    <a
      className="link row gridLink"
      href="/2"
      onClick={[Function]}
    >
      <h2>
        #2
      </h2>
      <p>
        Ethereum
      </p>
    </a>
  </li>
</ul>
`;

exports[`generateRoutes generates routes correctly 1`] = `null`;
